engine/engine.go

```diff
@@ -35,6 +35,9 @@ type Opts struct {
 	// DisableOptimizers disables Query optimizations using logicalPlan.DefaultOptimizers.
 	DisableOptimizers bool
 
+	// AdditionalOptimizers are optimizers which are run in addition to the default ones.
+	AdditionalOptimizers []logicalplan.Optimizer
```

Comment:
I would vote to put everything under the `DisableOptimizers` flag. I would guess that a use case like [this](https://github.com/prometheus/prometheus/issues/8053#issuecomment-708677714) is probably rare. At least I haven't seen anything like this in many years of using Prometheus. Another thing is that that use case is really relevant only for remote read in Prometheus i.e. when Prometheus retrieves data from other Prometheus nodes. In the case of Thanos, that can only be hit when push-down happens, when we execute a given query directly on Prometheus. That would be my 2 cents :smile:

Reply:
Yeah I think this one should be safe. But keeping it as an array rather than a Boolean should be fine since the engine is only used as a library? If Cortex/Mimir/Thanos doesn't want to enable that they can just expose that as flag or disable it directly

Reply:
👍 sounds good.

Reply:
Hm so I did not fully understand the conclusion. Do we want to have an array of optimizers, or do we want to have a boolean flag for "optional" optimizers?

Reply:
As far as I understood it, there should be an array internally so that library users could selectively enable/disable an optimizer but only a boolean flag should be exposed to end-users of software that uses the library like Thanos.

Reply:
Is there some value in allowing users to inject optimizers?

Reply:
I think it will be useful for promql-engine developers to be able to selectively enable/disable optimizers e.g. when debugging something. We always add the default ones [here](https://github.com/thanos-community/promql-engine/pull/115/files#diff-2e6c4934f63ff9b712c2c346b33036af4724adf70b0801fff9b74f71b37fcd89R140) so I think the current version is OK, we are still able to disable what is not needed when debugging :smile:

Reply:
@fpetkovski ping :point_up:

Reply:
Ah missed this. So if logical optimizers is empty, do we just use the default optimizers?

Reply:
I think so :smile: